File: c/image.c
// Demonstrates drawing pixels and the effect of glDrawPixels(), glCopyPixels(), and glPixelZoom().
// Moving the mouse while pressing the mouse button will copy the image in the lower-left corner of the window
// to the mouse position, using the current pixel zoom factors.
// There is no attempt to prevent you from drawing over the original image.
// If you press the 'z' or 'Z' keys, you change the zoom factors.
// If you press the 'r' key, the original image and zoom factors are reset.

#include <GL/glut.h>
#include <stdlib.h>
#include <stdio.h>

// Create checkerboard image.
#define	checkImageWidth 64
#define	checkImageHeight 64
GLubyte checkImage[checkImageHeight][checkImageWidth][3];

static GLdouble zoomFactor = 1.0;
static GLint height;

void makeCheckImage(void)
{
	int i, j, c;

	for(i = 0; i < checkImageHeight; i++)
	{
		for(j = 0; j < checkImageWidth; j++)
		{
			c = ( ((i&0x8)==0 ) ^ ((j&0x8)==0) )*255;
			checkImage[i][j][0] = (GLubyte) c;
			checkImage[i][j][1] = (GLubyte) c;
			checkImage[i][j][2] = (GLubyte) c;
		}
	}
}

void init(void)
{
	glClearColor(0.0, 0.0, 0.0, 0.0);
	glShadeModel(GL_FLAT);
	makeCheckImage();
	glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
}

void display(void)
{
	glClear(GL_COLOR_BUFFER_BIT);
	glRasterPos2i(0, 0);
	glDrawPixels(checkImageWidth, checkImageHeight, GL_RGB,
				 GL_UNSIGNED_BYTE, checkImage);
	glFlush();
}

void reshape(int w, int h)
{
	glViewport(0, 0, (GLsizei) w, (GLsizei) h);
	height = (GLint) h;
	glMatrixMode(GL_PROJECTION);
	glLoadIdentity();
	gluOrtho2D(0.0, (GLdouble) w, 0.0, (GLdouble) h);
	glMatrixMode(GL_MODELVIEW);
	glLoadIdentity();
}

void motion(int x, int y)
{
	static GLint screeny;
	screeny = height - (GLint) y;
	glRasterPos2i(x, screeny);
	glPixelZoom(zoomFactor, zoomFactor);
	glCopyPixels(0, 0, checkImageWidth, checkImageHeight, GL_COLOR);
	glPixelZoom(1.0, 1.0);
	glFlush();
}

void keyboard(unsigned char key, int x, int y)
{
	switch(key)
	{
		case 'r':
		case 'R':
			zoomFactor = 1.0;
			glutPostRedisplay();
			printf("zoomFactor reset to 1.0\n");
			break;

		case 'z':
			zoomFactor += 0.5;

			if(zoomFactor >= 3.0)
			{
				zoomFactor = 3.0;
			}

			printf("zoomFactor is now %4.1f\n", zoomFactor);
			break;

		case 'Z':
			zoomFactor -= 0.5;

			if(zoomFactor <= 0.5)
			{
				zoomFactor = 0.5;
			}

			printf("zoomFactor is now %4.1f\n", zoomFactor);
			break;

		case 27:
			exit(0);
			break;

		default:
			break;
	}
}

int main(int argc, char** argv)
{
	glutInit(&argc, argv);
	glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
	glutInitWindowSize(250, 250);
	glutInitWindowPosition(100, 100);
	glutCreateWindow(argv[0]);
	init();
	glutDisplayFunc(display);
	glutReshapeFunc(reshape);
	glutKeyboardFunc(keyboard);
	glutMotionFunc(motion);
	glutMainLoop();
	return 0;
}

File: python/Image.py
#!/usr/bin/python

from Window import *

def makeCheckImage():
	'''Create checkerboard texture.'''
	width  = 64
	height = 64
	checkImage = height*[width*[0]]
	for i in range(height):
		for j in range(width):
			c = ( ((i&0x8)==0 ) ^ ((j&0x8)==0) )*0xFF
			checkImage[i][j] = (c<<24)|(c<<16)|(c<<8)|(0xFF)
	return checkImage

class Image(Window):
	''' Demonstrates drawing pixels and the effect of glDrawPixels(), glCopyPixels(), and glPixelZoom().'''
	''' There is no attempt to prevent you from drawing over the original image.'''

	def __init__(self):
		super(Image, self).__init__("image.c", "Image", 250, 250)
		self.keybindings['z'] = self.zoomIn
		self.keybindings['Z'] = self.zoomOut
		self.keybindings['r'] = self.zoomReset
		self.zoomFactor = 1
		self.checkImage = makeCheckImage()
		glPixelStorei(GL_UNPACK_ALIGNMENT, 1)

	def display(self):
		glClear(GL_COLOR_BUFFER_BIT)
		glRasterPos2i(0, 0)
		glDrawPixels(64, 64, GL_RGB, GL_UNSIGNED_BYTE, self.checkImage)
		glFlush()

	def zoomReset(self):
		''' If you press the 'r' key, the original image and zoom factors are reset.'''
		self.zoomFactor = 1
		glutPostRedisplay()
		print "zoomFactor reset to 1"

	def zoomIn(self):
		''' If you press the 'z' or 'Z' keys, you change the zoom factors.'''
		self.zoomFactor += 0.5
		if(self.zoomFactor >= 3):
			self.zoomFactor = 3
		print "zoomFactor is now", self.zoomFactor

	def zoomOut(self):
		''' If you press the 'z' or 'Z' keys, you change the zoom factors.'''
		self.zoomFactor -= 0.5
		if(self.zoomFactor <= 0.5):
			self.zoomFactor = 0.5
		print "zoomFactor is now", self.zoomFactor

	def motion(self, x, y):
		''' Moving the mouse while pressing the mouse button will copy the image in the lower-left corner of the window'''
		''' to the mouse position, using the current pixel zoom factors.'''
		screeny = height - y
		glRasterPos2i(x, screeny)
		glPixelZoom(zoomFactor, zoomFactor)
		glCopyPixels(0, 0, checkImageWidth, checkImageHeight, GL_COLOR)
		glPixelZoom(1, 1)
		glFlush()

if __name__ == '__main__':
	Image().run()
